lib/internal/bootstrap/pre_execution.js

```diff
@@ -221,6 +221,51 @@ function initializeDeprecations() {
                                 'process.binding() is deprecated. ' +
                                 'Please use public APIs instead.', 'DEP0111');
   }
+
+  const { cjsContext } = require('internal/bootstrap/loaders');
+  const { getStackSourceName } = internalBinding('contextify');
+  const experimentalModules = getOptionValue('--experimental-modules');
+
+  // Deprecate global.process and global.Buffer access in ES Modules.
+  const processGetter = deprecate(
+    () => cjsContext.process,
+    'global.process is deprecated in ECMAScript Modules. ' +
+        'Please use `import process from \'process\'` instead.', 'DEP0XXX');
+  Object.defineProperty(global, 'process', {
+    get() {
+      if (experimentalModules) {
+        const sourceName = getStackSourceName(1);
+        if (sourceName && sourceName.startsWith('file://'))
+          return processGetter();
+      }
+      return cjsContext.process;
+    },
+    set(value) {
+      cjsContext.process = value;
+    },
+    enumerable: false,
+    configurable: true
+  });
+
+  const bufferGetter = deprecate(
+    () => cjsContext.Buffer,
+    'global.Buffer is deprecated in ECMAScript Modules. ' +
+        'Please use `import { Buffer } from \'buffer\'` instead.', 'DEP0YYY');
+  Object.defineProperty(global, 'Buffer', {
+    get() {
+      if (experimentalModules) {
+        const sourceName = getStackSourceName(1);
+        if (sourceName && sourceName.startsWith('file://'))
```

Comment:
This seems a bit brittle … are you sure there’s no better way to recognize ES modules? I think this also affects people who run scripts with `vm.runInContext(..., { filename: ... })` with a `file://` URL, at least.

Reply:
Yes this is the part I was discussing with v8 - I'll be driving a PR to use unique ScriptOrigin data here, so that this can be replaced in a few months hopefully.

Reply:
i don't feel comfortable with this landing while the check isn't correct. would it be possible to wait a little longer or apply a one-off patch to v8 for that feature in this case?

Reply:
I can appreciate that a string check doesn't have the uniqueness properties we want, and that is why I will immediately follow-up with a refined approach here in v8. But if we deem changing `global.process` from a value property to a getter property as semver major, then no, we cannot delay this.